x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/TopNOperator.java

```diff
@@ -273,6 +276,12 @@ public String describe() {
     private final List<SortOrder> sortOrders;
 
     private Row spare;
+    /**
+     * Esp. in case of many fields, a row's values can become quite large; growing from an empty row every time is expensive due
+     * to lots of allocations. Instead, keep track of the last row's actual size and use that to pre-allocate memory for the next row.
+     */
+    private int lastSeenRowValuesSize = 32;
```

Comment:
We could do the same for the row's search keys, but I'm not sure it's worth it.

For this to matter, there'd have to be a) many search keys with b) large values (e.g. sorting by 100 keyword fields). In this case, the bottleneck may be somewhere else entirely.

What do you think @nik9000 ?

Reply:
I'd do it. I guess I figure "doing it twice" doesn't really add an cognitive overhead.

One thing that's neat - we could report this in the "status" object and it'd tell us what the "biggest" row is.

Now - I suppose, a bigger problem might be - what happens if we have one row that's big and many tiny ones? I don't have a solution for this.

Reply:
I suppose a way to do this is to copy the built object into a second one that's perfectly sized if we're much too big.

Reply:
Can we maintain a `lastInsertedRow` and use it as a reference for the new row? We can pass this row as a parameter to the constructor of the Row. This way, we can initialize the keys and values with the sizes of the previous keys and values. I also think we should limit the initial sizes to avoid a big row with many tiny rows cases.

Reply:
Yeah, using the last row as indicator may be super-duper off, e.g. if we alternate between short and long rows - then we're always off. I'm not sure if using a `lastInsertedRow` would solve that. (The inserted rows sizes might not be indicative of the rows we still encounter and thus have to allocate for, but I may be wrong.)

I like the idea of copying out into a perfectly sized row. This way we can keep a single row as buffer, let it grow to whatever max size we need, and only discard it after the operator has run - this will minimize allocations even more.

Reply:
If we copied to a perfectly sized row I'd lower the max target - like half it or something. Or just set it back to the perfectly sized row. Then the next row is still perfectly sized. It'd work great for runs of big rows and runs of small rows. Much less good for big row/small row/big row/small row. But that feels less likely than runs.